### PR TITLE
Fix float uniform generation in TH

### PR DESCRIPTION
**Pull request by colesbury**
Generate random uniform floats in the range [0, 1) by generating random
uniform uint32 in the range [0, 2^24-1] and dividing by 2^24. This
ensures that the largest value is representable as a float32 less than
one.

This also changes the uniform double generation to use more bits of
randomness.

### CODE REVIEW

**Comment by colesbury** [comment]
```python
>>> torch.Tensor(100000000).uniform_().max()
0.9999999403953552
>>> import ctypes; 
>>> math = ctypes.cdll.LoadLibrary(None)
>>> math.nextafterf.restype = ctypes.c_float
>>> math.nextafterf(ctypes.c_float(1), ctypes.c_float(0))
0.9999999403953552
```

File: aten/src/TH/THRandom.c [hunk 1]
```diff
@@ -200,11 +200,33 @@ uint64_t THRandom_random(THGenerator *_generator)
...
+static double DOUBLE_DIVISOR = 1.0 / (1ULL << 53);
+
+// floats have 23 bits of mantissa (fractional part)
+static uint32_t FLOAT_MASK = (1 << 24) - 1;
+static float FLOAT_DIVISOR = 1.0f / (1 << 24);
+
+/* generates a random number on [0,1)-double-interval */
+static double uniform_double(THGenerator *_generator)
+{
+  uint64_t x = THRandom_random64(_generator);
```
Comment (colesbury) [inline aten/src/TH/THRandom.c]: NOTE: this now consumes two 32-bit random numbers (so calling functions may be slower), but should have more bits of randomness in the generated doubles. not sure if it's worth it in this change.

File: aten/src/TH/THRandom.c [hunk 2]
```diff
@@ -200,11 +200,33 @@ uint64_t THRandom_random(THGenerator *_generator)
   return y;
 }
 
+uint64_t THRandom_random64(THGenerator *_generator)
+{
+  uint64_t hi = THRandom_random(_generator);
```
Comment (soumith) [inline aten/src/TH/THRandom.c]: add a comment / TODO here that THRandom_random is only generating 32-bit of randomness

**Review by soumith — APPROVED** [review]
add the TODO (see inline comment), then it's good to go

**Comment by colesbury** [comment]
See #2740

Still need to fix CUDA uniform

**Review by apaszke — APPROVED** [review]

**Pull request merged**